### PR TITLE
FINERACT-1587: Add email field to Client

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

Allow add contact details to Client entity

[Apache Fineract JIRA ticket FINERACT-1587](https://issues.apache.org/jira/browse/FINERACT-1587)

## Related issues and discussion
#{Issue Number}

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] If you have multiple commits please combine them into one commit by squashing them.

- [ ] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/clients/client-stepper/client-general-step/client-general-step.component.ts [hunk 1]
```diff
@@ -76,6 +76,7 @@ export class ClientGeneralStepComponent implements OnInit {
       'externalId': [''],
       'genderId': [''],
       'mobileNo': [''],
+      'emailAddress': ['', Validators.email],
```
Comment (galovics) [inline src/app/clients/client-stepper/client-general-step/client-general-step.component.ts:79]: Since the email address should be a plain text, we shouldn't introduce validations to it. Am I misunderstanding the purpose of this?

File: src/app/clients/clients-routing.module.ts [hunk 2]
```diff
@@ -83,8 +83,7 @@ const routes: Routes = [
             data: { title: extract('General'), breadcrumb: 'General', routeParamBreadcrumb: false },
             resolve: {
               clientAccountsData: ClientAccountsResolver,
-              clientChargesData: ClientChargesResolver,
-              clientSummary: ClientSummaryResolver
```
Comment (galovics) [inline src/app/clients/clients-routing.module.ts:87]: Why removing the clientSummary?

File: src/app/clients/clients-view/general-tab/general-tab.component.ts [hunk 3]
```diff
@@ -73,7 +73,7 @@ export class GeneralTabComponent {
       this.loanAccounts = data.clientAccountsData.loanAccounts;
       this.shareAccounts = data.clientAccountsData.shareAccounts;
       this.upcomingCharges = data.clientChargesData.pageItems;
-      this.clientSummary = data.clientSummary[0];
+      this.clientSummary = data.clientSummary ? data.clientSummary[0] : [];
```
Comment (galovics) [inline src/app/clients/clients-view/general-tab/general-tab.component.ts:76]: Is this a bugfix?

File: src/app/clients/edit-client/edit-client.component.html [hunk 4]
```diff
@@ -84,9 +84,21 @@
             <input matInput formControlName="externalId">
           </mat-form-field>
 
+          <mat-checkbox *ngIf="editClientForm.value.legalFormId === 1" fxFlex="48%" labelPosition="before" formControlName="isStaff" class="margin-v">
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:87]: Is this related to the email field?

File: src/app/clients/edit-client/edit-client.component.html [hunk 5]
```diff
@@ -109,20 +121,20 @@
 
           <mat-form-field fxFlex="48%">
             <mat-label>Submitted On</mat-label>
-            <input matInput [min]="minDate" [max]="maxDate" [matDatepicker]="submittedOnDatePicker" formControlName="submittedOnDate">
+            <input matInput [min]="minDate" [max]="maxDate" [matDatepicker]="submittedOnDatePicker" formControlName="submittedOnDate" required>
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:124]: This is not related to the email field, is it?

File: src/app/clients/edit-client/edit-client.component.html [hunk 6]
```diff
@@ -109,20 +121,20 @@
...
             <mat-datepicker-toggle matSuffix [for]="submittedOnDatePicker"></mat-datepicker-toggle>
             <mat-datepicker #submittedOnDatePicker></mat-datepicker>
+            <mat-error *ngIf="editClientForm.controls.submittedOnDate.hasError('required')">
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:127]: Same for this.

File: src/app/clients/edit-client/edit-client.component.html [hunk 7]
```diff
@@ -109,20 +121,20 @@
...
+              Submitted Date is <strong>required</strong>
+            </mat-error>
           </mat-form-field>
 
           <mat-form-field fxFlex="48%">
             <mat-label>Activated On</mat-label>
-            <input matInput [min]="minDate" [max]="maxDate" [matDatepicker]="activatedOnDatePicker" required
+            <input matInput [min]="minDate" [max]="maxDate" [matDatepicker]="activatedOnDatePicker"
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:134]: Same here.

File: src/app/clients/edit-client/edit-client.component.html [hunk 8]
```diff
@@ -109,20 +121,20 @@
...
               formControlName="activationDate">
             <mat-datepicker-toggle matSuffix [for]="activatedOnDatePicker"></mat-datepicker-toggle>
             <mat-datepicker #activatedOnDatePicker></mat-datepicker>
-            <mat-error *ngIf="editClientForm.controls.activationDate.hasError('required')">
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:123]: Why removing?

File: src/app/clients/edit-client/edit-client.component.html [hunk 9]
```diff
@@ -134,10 +146,6 @@
             </mat-select>
           </mat-form-field>
 
-          <mat-checkbox *ngIf="editClientForm.value.legalFormId === 1" fxFlex="48%" labelPosition="before" formControlName="isStaff" class="margin-v">
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.html:137]: Has this been moved above? If so, why?

File: src/app/clients/edit-client/edit-client.component.ts [hunk 10]
```diff
@@ -101,11 +102,12 @@ export class EditClientComponent implements OnInit {
...
       'genderId': [''],
       'mobileNo': [''],
+      'emailAddress': ['', Validators.email],
       'dateOfBirth': [''],
       'clientTypeId': [''],
       'clientClassificationId': [''],
-      'submittedOnDate': [''],
-      'activationDate': ['', Validators.required]
+      'submittedOnDate': ['', Validators.required],
+      'activationDate': ['']
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.ts:110]: You sure this is related to the email field?

File: src/app/clients/edit-client/edit-client.component.ts [hunk 11]
```diff
@@ -101,11 +102,12 @@ export class EditClientComponent implements OnInit {
       'externalId': [''],
       'genderId': [''],
       'mobileNo': [''],
+      'emailAddress': ['', Validators.email],
```
Comment (galovics) [inline src/app/clients/edit-client/edit-client.component.ts:105]: Same as above for the validations, it's a plan text field, I don't think this is needed.

File: src/app/settings/settings.service.ts [hunk 12]
```diff
@@ -57,6 +64,9 @@ export class SettingsService {
    * Returns language setting
    */
   get language() {
+    if (!localStorage.getItem('mifosXLanguage')) {
```
Comment (galovics) [inline src/app/settings/settings.service.ts:67]: I assume this is a bugfix but then why didn't you create a separate PR?

File: src/app/shared/language-selector/language-selector.component.html [hunk 13]
```diff
@@ -1,6 +1,6 @@
 <mat-form-field id="language-selector">
   <mat-label>Language</mat-label>
-  <mat-select [formControl]="languageSelector">
+  <mat-select [formControl]="languageSelector" (selectionChange)="setLanguage()" class="languageselector">
```
Comment (galovics) [inline src/app/shared/language-selector/language-selector.component.html:3]: Bugfix?

**Review by galovics — COMMENTED** [review]
Couple of questions but mainly I'm really not sure if everything is related to the email field. Would be nice to introduce separate PRs for the different set of changes, bugfixes and for the email field.

File: src/app/clients/clients-routing.module.ts (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/clients-routing.module.ts:87]: Usually there is an error with that report that blocks other steps

File: src/app/clients/edit-client/edit-client.component.html (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/edit-client/edit-client.component.html:87]: It is just to set a better view distribution, I mean set the contact data, mobile and email in the same line

File: src/app/clients/edit-client/edit-client.component.html (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/edit-client/edit-client.component.html:127]: Same as above, The submitted date must be required in the Edit client view due this value was added previously

File: src/app/clients/edit-client/edit-client.component.html (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/edit-client/edit-client.component.html:134]: In the edit case, If the client has not been activated yet why this field must be required. It's clear when the client is already activated It must be required.

File: src/app/clients/edit-client/edit-client.component.html (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/edit-client/edit-client.component.html:123]: The same reason as above, If the client is not activated yet, It can't be edited If we keep the activation date as required

File: src/app/clients/edit-client/edit-client.component.html (hunk 9 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/edit-client/edit-client.component.html:137]: It is just to set a better view distribution, I mean set the contact data, mobile and email in the same line

File: src/app/shared/language-selector/language-selector.component.html (hunk 13 above)
Comment (josehernandezfintecheandomx) [reply src/app/shared/language-selector/language-selector.component.html:3]: Yes Language selector fix to start the support of i18n in this application

File: src/app/settings/settings.service.ts (hunk 12 above)
Comment (josehernandezfintecheandomx) [reply src/app/settings/settings.service.ts:67]: Related to the Language fix to start the support of i18n in this application

File: src/app/clients/client-stepper/client-general-step/client-general-step.component.ts (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply src/app/clients/client-stepper/client-general-step/client-general-step.component.ts:79]: Yes, you're right but I think that is better to set a basic email validation

**Review by galovics — APPROVED** [review]

**Comment by galovics** [comment]
As far as I see the linter fails, please fix that @josehernandezfintecheandomx and then you're good to go.
